server/app/views/components/FieldWithLabel.java

```diff
@@ -246,107 +243,183 @@ public FieldWithLabel showFieldErrors(boolean showFieldErrors) {
-      // Setting inputmode to decimal gives iOS users a more accessible keyboard
-      fieldTag.attr("inputmode", "decimal");
-
-      // Setting step to any disables the built-in HTML validation so we can use our
-      // custom validation message to enforce integers.
-      fieldTag.attr("step", "any");
-
-      // Set min and max values for client-side validation
-      if (this.minValue.isPresent()) {
-        fieldTag.attr("min", minValue.getAsLong());
-      }
-      if (this.maxValue.isPresent()) {
-        fieldTag.attr("max", maxValue.getAsLong());
-      }
-
-      // For number types, only set the value if it's present since there is no empty string
-      // equivalent for numbers.
-      if (this.fieldValueNumber.isPresent()) {
-        fieldTag.withValue(String.valueOf(this.fieldValueNumber.getAsLong()));
-      }
-    } else {
-      fieldTag.withValue(this.fieldValue);
+  }
+
+  protected LabelTag genLabelTag() {
+
+    return label()
+        .attr(Attr.FOR, this.id)
+        // If the text is screen-reader text, then we want the label to be screen-reader
+        // only.
+        .withClass(labelText.isEmpty() ? Styles.SR_ONLY : BaseStyles.INPUT_LABEL)
+        .withText(labelText.isEmpty() ? screenReaderText : labelText);
+  }
+
+  protected DivTag wrapInDivTag(Tag fieldTag, Tag labelTag, String fieldErrorsId) {
+    return div(
+            labelTag,
+            div(fieldTag, buildFieldErrorsTag(fieldErrorsId))
+                .withClasses(Styles.FLEX, Styles.FLEX_COL))
+        .withClasses(
+            StyleUtils.joinStyles(referenceClassesBuilder.build().toArray(new String[0])),
+            BaseStyles.FORM_FIELD_MARGIN_BOTTOM);
+  }
+
+  protected boolean getHasFieldErrors() {
+    return !fieldErrors.isEmpty() && showFieldErrors;
+  }
+
+  protected void numberTagApplyAttrs(Tag fieldTag) {
+    // Setting inputmode to decimal gives iOS users a more accessible keyboard
+    fieldTag.attr("inputmode", "decimal");
+
+    // Setting step to any disables the built-in HTML validation so we can use our
+    // custom validation message to enforce integers.
+    fieldTag.attr("step", "any");
+
+    // Set min and max values for client-side validation
+    if (this.minValue.isPresent()) {
+      fieldTag.attr("min", minValue.getAsLong());
+    }
+    if (this.maxValue.isPresent()) {
+      fieldTag.attr("max", maxValue.getAsLong());
     }
 
-    boolean hasFieldErrors = !fieldErrors.isEmpty() && showFieldErrors;
-    if (hasFieldErrors) {
-      fieldTag.attr("aria-invalid", "true");
-      fieldTag.attr("aria-describedBy", fieldErrorsId);
+    // For number types, only set the value if it's present since there is no empty string
+    // equivalent for numbers.
+    if (this.fieldValueNumber.isPresent()) {
+      fieldTag.attr("value", String.valueOf(this.fieldValueNumber.getAsLong()));
     }
+  }
 
+  protected void generalApplyAttrsClassesToTag(Tag fieldTag, boolean hasFieldErrors) {
```

Comment:
Can this take a more specific type? If we're setting the "name" and "placeholder" attributes, we should ensure that the specified tag passed in supports them.

Reply:
As is, this method takes in an `InputTag`, a `TextareaTag`, and in the subclass a `SelectTag`

The `.withClasses`, `.withId`, and `.condAttr` are from `Tag` while the `.withName` is from the `IName` interface

I think I found a way to make it work with all three types being passed to it

What do you think? Although its not entirely 'specific' but it is a bit more I guess?